identity: Reduce churn, fix manager refcounts

jrajahalme opened the pull request:
Use identitymanager.Add instead of identitymanager.RemoveOldAddNew
when initially adding the identity of a new endpoint into the identity
manager. This fixes a refcounting bug, where upon the restoration of
two endpoints with the same labels the first restoration adds the
identity to the manager with refcount 1, but the second restoration
first deletes the lone refcount, triggering removal of the associated
policy from the policy cache, and then adds the same ID back, with the
resulting refcount of 1, even when two endpoints are using the same
ID.

Require the old identity given to RemoveOldAddNew to have been added
to the identity manager previously, and make the case where
RemoveOldAddNew is called with two identities with the same numerical
ID a no-op to reduce churn in the observers.

Fixes: #8685
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8691)
<!-- Reviewable:end -->

joestringer commented on daemon/state.go:
```diff
@@ -353,7 +353,7 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 			// parts of the identity not being marshaled to JSON. Hence we must set
 			// the identity even if has not changed.
-			ep.SetIdentity(identity)
+			ep.SetIdentity(identity, true)
```
We might have already called `SetIdentity(oldIdentity, true`) in the `ParseEndpoint()` path in `pkg/endpoint`, then set a new, different identity here (because of different `--labels` options on the daemon). In that case we should actually remove a reference to the old identity.

jrajahalme replied:
Thanks, did not notice that. Planning to make `ParseEndpoint()` NOT call `SetIdentity()` at all, but ensure that `ep.SecurityIdentity` is non-nil:
```
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 2dde1a4c89..c560343f28 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -1086,11 +1086,12 @@ func ParseEndpoint(owner regeneration.Owner, strEp string) (*Endpoint, error) {
 		ep.Status = NewEndpointStatus()
 	}
 
+	// Make sure the endpoint has an identity, using the 'init' identity if none.
 	if ep.SecurityIdentity == nil {
-		ep.SetIdentity(identityPkg.LookupReservedIdentity(identityPkg.ReservedIdentityInit), true)
-	} else {
-		ep.SecurityIdentity.Sanitize()
+		ep.SecurityIdentity = identityPkg.LookupReservedIdentity(identityPkg.ReservedIdentityInit)
 	}
+	ep.SecurityIdentity.Sanitize()
+
 	ep.UpdateLogger(nil)
 
 	ep.SetStateLocked(StateRestoring, "Endpoint restoring")
```

This way the state of the endpoint's identity is consistent after `ParseIdentity()`: not registered with the identitymanager yet.

jrajahalme committed:
identity: Fix manager refcounts, reduce churn

Use identitymanager.Add instead of identitymanager.RemoveOldAddNew
when initially adding the identity of a new endpoint into the identity
manager. This fixes a refcounting bug, where upon the restoration of
two endpoints with the same labels the first restoration adds the
identity to the manager with refcount 1, but the second restoration
first deletes the lone refcount, triggering removal of the associated
policy from the policy cache, and then adds the same ID back, with the
resulting refcount of 1, even when two endpoints are using the same
ID.

Require the old identity given to RemoveOldAddNew to have been added
to the identity manager previously, and make the case where
RemoveOldAddNew is called with two identities with the same numerical
ID a no-op to reduce churn in the observers.

ParseEndpoint(), which is used to restore endpoints on restart, used
to call SetIdentity() to set a nil security identity to the
init-identity.  This is changed so that the caller is responsible for
calling SetIdentity() in all cases.

Fixes: #8685
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
